libraries/lib-snapping/ProjectSnap.cpp

```diff
@@ -0,0 +1,120 @@
+#include "ProjectSnap.h"
+#include "Project.h"
+#include "XMLWriter.h"
+#include "XMLAttributeValueView.h"
+
+#include "ProjectRate.h"
+#include "ProjectTimeSignature.h"
+
+static const AudacityProject::AttachedObjects::RegisteredFactory sKey {
+   [](AudacityProject& project)
+   {
+      auto result = std::make_shared<ProjectSnap>(project);
+      return result;
+   }
+};
+
+ProjectSnap& ProjectSnap::Get(AudacityProject& project)
+{
+   return project.AttachedObjects::Get<ProjectSnap>(sKey);
+}
+
+const ProjectSnap& ProjectSnap::Get(const AudacityProject& project)
+{
+   return Get(const_cast<AudacityProject&>(project));
+}
+
+ProjectSnap::ProjectSnap(const AudacityProject& project)
+    : mProject(project)
+{}
+
+void ProjectSnap::SetSnapMode(SnapMode mode)
+{
+   if (mSnapMode != mode)
+   {
+      mSnapMode = mode;
+      
+      SnapModeSetting.WriteEnum(mSnapMode);
+      gPrefs->Flush();
+
+      Publish(SnapChangedMessage { mode, mSnapTo });
+   }
+}
+
+SnapMode ProjectSnap::GetSnapMode() const
+{
+   return mSnapMode;
+}
+
+void ProjectSnap::SetSnapTo(Identifier snap)
+{
+   if (mSnapTo != snap)
+   {
+      mSnapTo = snap;
+
+      SnapToSetting.Write(mSnapTo.GET());
+      gPrefs->Flush();
+
+      Publish(SnapChangedMessage { mSnapMode, snap });
+   }
+}
+
+Identifier ProjectSnap::GetSnapTo() const
+{
+   return mSnapTo;
+}
+
+SnapResult ProjectSnap::SnapTime(double time) const
+{
+   if (mSnapMode == SnapMode::SNAP_OFF)
+      return { time, false };
+   
+   SnapConfig config {
+      ProjectRate::Get(mProject).GetRate(),
+      ProjectTimeSignature::Get(mProject).GetTempo(),
+      { ProjectTimeSignature::Get(mProject).GetUpperTimeSignature(),
+        ProjectTimeSignature::Get(mProject).GetLowerTimeSignature() }
+   };
+   
+   return SnapFunctionsRegistry::Snap(mSnapTo, config, time, mSnapMode == SnapMode::SNAP_NEAREST);
+}
+
+static ProjectFileIORegistry::AttributeWriterEntry entry {
+   [](const AudacityProject& project, XMLWriter& xmlFile)
+   {
+      // Keep this attr for compatibility
+      auto& snapSettings = ProjectSnap::Get(project);
+      xmlFile.WriteAttr(
+         wxT("snapto"), snapSettings.GetSnapMode() != SnapMode::SNAP_OFF ?
+                           wxT("on") :
+                           wxT("off"));
+   }
+};
+
+static ProjectFileIORegistry::AttributeReaderEntries entries {
+   // Just a pointer to function, but needing overload resolution as non-const:
+   (ProjectSnap & (*)(AudacityProject&)) & ProjectSnap::Get,
+   {
+      // PRL:  The following has persisted as a per-project setting for long.
+      // Maybe that should be abandoned.  Enough to save changes in the user
+      // preference file.
+      // DV: Keep the old behavior here: project only stored *off* and *nearest*
+      { "snapto",
+        [](auto& snapSettings, auto value)
+        {
+           snapSettings.SetSnapMode(
+              value.ToWString() == wxT("on") ? SnapMode::SNAP_NEAREST :
```

Comment:
So the distinction between nearest and prior is lost after saving and reloading the project.  Should we try to fix that?

Reply:
This is the original behavior